### DISCUSSION TITLE
error occured in SAS during sasdata2dataframe

### DISCUSSION THREAD

**Discussion by DionisMotor** [sassoftware/saspy#565] [Q&A] [answered]
Hi

Just starting with sapy. Sorry, but can't solve this by myself.
Actually I had an issue already from a start.

`cars_dataframe = sas.sd2df('cars', 'sashelp')`

and a log looks like:

> ...
> 110  filename sock socket 'localhost:52797' recfm=s encoding='utf-8' lrecl=1048576;
> ERROR: Could not load /opt/sas/stratum/prog/SASFoundation/9.4/dbcs/sasexe/sasxbamt (80 images loaded)
> ERROR: libnsl.so.1: cannot open shared object file: No such file or directory
> ERROR: Invalid logical name.
> ERROR: Error in the FILENAME statement.
> 111  data _null_; set sashelp.'cars'n ;
> ...

mentioned file exist:
> -r-xr-xr-x 1 sasadmin sassrv 259247 Aug  6  2020 /opt/sas/stratum/prog/SASFoundation/9.4/dbcs/sasexe/sasxbamt

What should I do ? 
**SAS version:**
> For Base SAS Software ...
>    Custom version information: 9.4_M7
>    Image version information: 9.04.01M7P080520

**Python 3.8.16**

Thanks
Denys

**Comment by tomweber-sas** [comment, marked as answer]
Hey, sorry to hear that, but I can help. This isn't a saspy thing, but rather a SAS/ Operating System requirement issue. The sasxbamt module isn't the one that's not there, it's part of the SAS deployment. But, it has OS level executable requirements that aren't part of the SAS deployment. That version of SAS was built and linked with certain OS version requirements, which is the issue here. It expects there to be this file `libnsl.so.1` available at the OS level, but it is not being found. there's probably a different version of that library on your system, but not that version.

So this is a systems requirement between the version of SAS you have installed and the version of the OS its installed on. The best thing to do would be to check w/ your systems admin (SAS admin) folks to see if they can make the OS level comply with what SAS needs. This may be just one of the same kind of issue that could be hit running other SAS code that loads other modules which rely on specific versions of the OS libraries. Also, checking with our Tech Support on how to make this compliant is another thing too. 

For reference, on my system (you's may be different) I have that library in the following location, with more than one version. 
```
tom64-7> locate libnsl.so
/usr/lib64/libnsl.so.1
/usr/lib64/libnsl.so.2
/usr/lib64/libnsl.so.2.0.0
tom64-7>
```
It may be that simply adding that version of the library, in to the same path where you have a different version, is the answer but that begs the question of is the OS isn't at the compliant level for the version of SAS, then what's the next one you're going to hit ...

But, maybe that'll get you going. Either way, this is really a systems admin exercise, not really an end users. So, I'll let you decide what's the right way for you to try to resolve this at your site. But that's what the problem is!

Tom

**Reply by DionisMotor** [reply]
Hi Tom
Thanks for answer.

Yep, on my system I have only 
```
/usr/lib64/libnsl.so.2
/usr/lib64/libnsl.so.2.0.0
```
I will try reach our admins. 

BR,
Denys

**Comment by MrShiv** [comment]
I realize I'm a bit late to this thread, but here's a quick fix: Add a symlink to libnsl.so V2 in lib64. 

`$ cd /usr/lib64/`
`$ sudo ln -s libnsl.so.2.0.0 libnsl.so.1`

You can't just copy libnsl.so.1 over from /usr/lib/, or you'll get the SAS error "libnsl.so.1: wrong ELF class: ELFCLASS32"
